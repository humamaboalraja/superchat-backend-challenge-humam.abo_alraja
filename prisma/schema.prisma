datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model Contact {
  id           String    @id @default(cuid())
  name         String
  email        String    @unique
  messagesFrom Message[] @relation("sender")
  messagesTo   Message[] @relation("reciever")
  createdAt    DateTime  @default(now())
  updatedAt    DateTime  @updatedAt
}

model Message {
  id             String   @id @default(cuid())
  messageContent String
  senderId       String
  recieverId     String   @unique
  sender         Contact  @relation(name: "sender", fields: [senderId], references: [id])
  reciever       Contact  @relation(name: "reciever", fields: [recieverId], references: [id])
  createdAt      DateTime @default(now())
  updatedAt      DateTime @updatedAt
}
