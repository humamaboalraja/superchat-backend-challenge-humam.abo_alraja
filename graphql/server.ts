import { ApolloServer } from 'apollo-server-express';
import express from 'express';
import { createServer } from 'http';
import { createContext } from './context';
import { schema } from './schema';

const { PORT = 5000 } = process.env;

const app = express();

const server = createServer(app);

export const apollo = new ApolloServer({
  schema,
  introspection: process.env.STAGE !== 'production',
  apollo: {},
  context: createContext,
});

export async function main() {
  await apollo.start();
  apollo.applyMiddleware({ app });
  server.listen({ port: PORT }, () => {
    process.stdout.write(
      ` ðŸš€  Server ready at http://localhost:${PORT}${apollo.graphqlPath}`
    );
  });
}
process.env.STAGE != 'test' && main();
