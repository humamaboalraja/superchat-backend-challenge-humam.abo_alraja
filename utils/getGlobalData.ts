import { prisma } from '../singletons/prisma';

export default async function getGlobalData() {
  const messages = await prisma.message.findMany({});
  const contacts = await prisma.contact.findMany({});
  if (!messages || !contacts) throw new Error('Unable to set up global data');
  return { messages, contacts };
}
