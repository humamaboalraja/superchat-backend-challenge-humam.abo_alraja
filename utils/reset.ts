import { prisma } from '../singletons/prisma';

export default async function reset() {
  try {
    const models: string[] = ['Message', 'Contact'];
    for (const model in models) {
      await prisma.$executeRaw(`DELETE FROM "${models[model]}";`);
    }
  } catch (e) {
    throw new Error(`${e}`);
  }
}
