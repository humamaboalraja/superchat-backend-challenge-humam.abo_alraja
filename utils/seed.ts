import cryptoRandomString from 'crypto-random-string';
import { prisma } from '../singletons/prisma';

export default async function seed() {
  try {
    const contact1 = await prisma.contact.create({
      data: {
        name: 'Humam Abo Alraja',
        email: 'humam@company.com',
      },
    });
    const contact2 = await prisma.contact.create({
      data: {
        name: 'John',
        email: 'john@company.com',
      },
    });

    const contact3 = await prisma.contact.create({
      data: {
        name: 'Emma Mockassin',
        email: 'emma@company.com',
      },
    });

    const contact4 = await prisma.contact.create({
      data: {
        name: 'Alex Smith',
        email: 'alex@company.com',
      },
    });

    if (!contact1 || !contact2 || !contact3 || !contact4)
      throw new Error('no contacts');

    const one = await prisma.message.create({
      data: {
        messageContent: 'Hey Emma, How is it going?!',
        sender: {
          connect: {
            id: contact1.id,
          },
        },
        reciever: {
          connect: {
            id: contact3.id,
          },
        },
      },
    });

    const two = await prisma.message.create({
      data: {
        messageContent: "Hi Humam, it's going great! How about you?",
        sender: {
          connect: {
            id: contact3.id,
          },
        },
        reciever: {
          connect: {
            id: contact1.id,
          },
        },
      },
    });

    const three = await prisma.message.create({
      data: {
        messageContent:
          'Hey Alex, do you mind sending me the pitch ASAP? Thanks in advance!',
        sender: {
          connect: {
            id: contact3.id,
          },
        },
        reciever: {
          connect: {
            id: contact4.id,
          },
        },
      },
    });

    const four = await prisma.message.create({
      data: {
        messageContent: 'Hi Emma, Definitely!',
        sender: {
          connect: {
            id: contact4.id,
          },
        },
        reciever: {
          connect: {
            id: contact3.id,
          },
        },
      },
    });
  } catch (e) {
    throw new Error(`${e}`);
  }
}
