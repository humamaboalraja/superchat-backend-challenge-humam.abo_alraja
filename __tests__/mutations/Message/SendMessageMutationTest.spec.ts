import { Message } from '@prisma/client';
import { prisma } from '../../../singletons/prisma';
import getGlobalData from '../../../utils/database/getGlobalData';
import reset from '../../../utils/database/reset';
import seed from '../../../utils/database/seed';
import server from '../../server';
describe('Sending a message mutations Tests', () => {
  beforeAll(async () => {
    await reset();
    await seed();
    global.testData = await getGlobalData();
  });

  const SEND_MESSAGE_MUTATION = `
  mutation SendMessage($inputType: SendMessageInput) {
    sendMessage(InputType: $inputType) {
      senderId
      recieverId
      messageContent
    }
  }`;

  const exampleArgs = {
    senderId: 'ckwcm9hzq0007y5hfzz0g4tcj',
    recieverId: 'ckwcm9hzf0000y5hfp8p74h7e',
    messageContent: 'Placeholderfff',
  };

  const payload = {
    data: {
      senderId: 'ckwcm9hzq0007y5hfzz0g4tcj',
      recieverId: 'ckwcm9hzf0000y5hfp8p74h7e',
      messageContent: 'Placeholderfff',
    },
  };

  let message: Message[];
  it('should send a message', async () => {
    // message = await prisma.message.create(payload);

    // TODO - Creating two contacts
    // TODO - Sending messages between contacts
    // TODO - Retrieving sent messages

    const res = await server.executeOperation({
      query: SEND_MESSAGE_MUTATION,
      variables: {
        inputType: {
          ...exampleArgs,
          senderId: global.testData.contacts[1].id,
          recieverId: global.testData.contacts[2].id,
          messageContent: 'sdfadsfs',
        },
      },
    });
    const message = await prisma.message.findMany({
      where: {
        senderId: 'ckwcm9hzf0000y5hfp8p74h7e',
        recieverId: 'ckwcm9hzf0000y5hfp8p74h7e',
      },
    });
    if (!message) throw new Error('Error fetching the sent message');
    expect(res).toMatchSnapshot();
  });
});

export {};
