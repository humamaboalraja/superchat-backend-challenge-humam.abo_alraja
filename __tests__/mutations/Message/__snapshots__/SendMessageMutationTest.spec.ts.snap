// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sending a message mutations Tests should send a message 1`] = `
Object {
  "data": Object {
    "sendMessage": Object {
      "messageContent": "sdfadsfs",
      "recieverId": "ckwdeztmj00141chf6rqlxv8f",
      "senderId": "ckwdeztm900071chfsi1fz8hv",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
